examples/lottery/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+#![no_std]
+
+imports!();
+
+const PERCENTAGE_TOTAL: u16 = 100;
+const THIRTY_DAYS_IN_SECONDS: u64 = 60 * 60 * 24 * 30;
+
+#[elrond_wasm_derive::contract(LotteryImpl)]
+pub trait Lottery {
+    
+    #[init]
+    fn init(&self) {
+        
+    }
+
+    #[endpoint]
+    fn start(&self,
+        lottery_name: Vec<u8>,
+        ticket_price: BigUint, 
+        total_tickets: Option<u32>, 
+        deadline: Option<u64>,
+        max_entries_per_user: Option<u32>,
+        prize_distribution: Option<Vec<u8>>,
+        whitelist: Option<Vec<Address>>) 
+        -> SCResult<()> {
+        
+        let timestamp = self.get_block_timestamp();
+        
+        let tt = total_tickets.unwrap_or(u32::MAX);
+        let d = deadline.unwrap_or(timestamp + THIRTY_DAYS_IN_SECONDS);
+        let max = max_entries_per_user.unwrap_or(u32::MAX);
+        let pd = prize_distribution.unwrap_or([PERCENTAGE_TOTAL as u8].to_vec());
+        let wl = whitelist.unwrap_or(Vec::new());
+
+        if self.status(lottery_name.clone()) != Status::Inactive {
+            return sc_error!("Lottery is already active!");
+        }
+        if ticket_price == 0 {
+            return sc_error!("Ticket price must be higher than 0!");
+        }
+        if tt == 0 {
+            return sc_error!("Must have more than 0 tickets available!");
+        }
+        if d <= timestamp {
+            return sc_error!("Deadline can't be in the past!");
+        }
+        if d > timestamp + THIRTY_DAYS_IN_SECONDS {
+            return sc_error!("Deadline can't be later than 30 days from now!");
+        }
+        if max == 0 {
+            return sc_error!("Must have more than 0 max entries per user!");
+        }
+        {
+            let mut sum = 0u16; // u16 to protect against overflow
+
+            for i in 0..pd.len() {
+                sum += pd[i] as u16;
+
+                if sum > PERCENTAGE_TOTAL {
+                    return sc_error!("Prize distribution must add up to exactly 100(%)!");
+                }
+            }
+            
+            if sum != PERCENTAGE_TOTAL {
+                return sc_error!("Prize distribution must add up to exactly 100(%)!");
+            }
+        }
+    
+        self.set_ticket_price(&lottery_name, ticket_price);
```

Comment:
instead of having 6 different sets, use a structure and write only once. Use serialization and deserialization for that.